software/contrib/pams.py

```diff
@@ -0,0 +1,1118 @@
+WAVE_START = 5
+
+## Gate, on while the clock is running and
+#  off when the clock stops
+WAVE_RUN = 6
+
+## Use raw AIN as the direct input
+#
+#  This lets you effectively use Pam's as a quantizer for
+#  the AIN signal
+WAVE_AIN = 7
+
+## Available wave shapes
+WAVE_SHAPES = [
+    WAVE_SQUARE,
+    WAVE_TRIANGLE,
+    WAVE_SIN,
+    WAVE_RANDOM,
+    WAVE_RESET,
+    WAVE_START,
+    WAVE_RUN,
+    WAVE_AIN
+]
+
+## Ordered list of labels for the wave shape chooser menu
+WAVE_SHAPE_LABELS = [
+    "Square",
+    "Triangle",
+    "Sine",
+    "Random",
+    "Reset",
+    "Start",
+    "Run",
+    "AIN"
+]
+
+## Sorted list of wave shapes to display
+#
+#  Same order as WAVE_SHAPE_LABELS
+#
+#  These are 12x12 bitmaps. See:
+#  - https://github.com/Allen-Synthesis/EuroPi/blob/main/software/oled_tips.md
+#  - https://github.com/novaspirit/img2bytearray
+WAVE_SHAPE_IMGS = [
+    bytearray(b'\xfe\x00\x82\x00\x82\x00\x82\x00\x82\x00\x82\x00\x82\x00\x82\x00\x82\x00\x82\x00\x82\x00\x83\xf0'),
+    bytearray(b'\x06\x00\x06\x00\t\x00\t\x00\x10\x80\x10\x80 @ @@ @ \x80\x10\x80\x10'),
+    bytearray(b'\x10\x00(\x00D\x00D\x00\x82\x00\x82\x00\x82\x10\x82\x10\x01\x10\x01\x10\x00\xa0\x00@'),
+    bytearray(b'\x00\x00\x08\x00\x08\x00\x14\x00\x16\x80\x16\xa0\x11\xa0Q\xf0Pp`P@\x10\x80\x00'),
+    bytearray(b'\x03\xf0\x02\x00\x02\x00\x02\x00\x02\x00\x02\x00\x02\x00\x02\x00\x02\x00\x02\x00\x02\x00\xfe\x00'),
+    bytearray(b'\xe0\x00\xa0\x00\xa0\x00\xa0\x00\xa0\x00\xa0\x00\xa0\x00\xa0\x00\xa0\x00\xa0\x00\xa0\x00\xbf\xf0'),
+    bytearray(b'\xff\xf0\x80\x00\x80\x00\x80\x00\x80\x00\x80\x00\x80\x00\x80\x00\x80\x00\x80\x00\x80\x00\x80\x00'),
+    bytearray(b'\x00\x00|\x00|\x00d\x00d\x00g\x80a\x80\xe1\xb0\xe1\xb0\x01\xf0\x00\x00\x00\x00')
+]
+
+STATUS_IMG_LOCK = bytearray(b'\x06\x00\x19\x80\x19\x80`@`@`@\xff\xf0\xf9\xf0\xf9\xf0\xfd\xf0\xff\xf0\xff\xf0')
+STATUS_IMG_PLAY = bytearray(b'\x00\x00\x18\x00\x18\x00\x1c\x00\x1c\x00\x1e\x00\x1f\x80\x1e\x00\x1e\x00\x1c\x00\x18\x00\x18\x00')
+STATUS_IMG_PAUSE = bytearray(b'\x00\x00y\xc0y\xc0y\xc0y\xc0y\xc0y\xc0y\xc0y\xc0y\xc0y\xc0y\xc0')
+
+STATUS_IMG_WIDTH = 12
+STATUS_IMG_HEIGHT = 12
+
+## Duration before we activate the screensaver
+SCREENSAVER_TIMEOUT_MS = 1000 * 60 * 5
+
+## Duration before we blank the screen
+BLANK_TIMEOUT_MS = 1000 * 60 * 20
+
+## Do we use gate input on din to turn the module on/off
+DIN_MODE_GATE = 'Gate'
+
+## Do we toggle the module on/off with a trigger on din?
+DIN_MODE_TRIGGER = 'Trig'
+
+DIN_MODES = [
+    DIN_MODE_GATE,
+    DIN_MODE_TRIGGER
+]
+
+class Setting:
+    """A single setting that can be loaded, saved, or dynamically read from an analog input
+    """
+    def __init__(self, display_name, storage_name, display_options, options, allow_cv_in=True, on_change_fn=None, value_dict=None, default_value=None):
+        self.display_name = display_name
+        self.display_options = []
+        for o in display_options:
+            self.display_options.append(o)
```

Comment:
I've always had an instinctive dislike of tricks like this that reduce line numbers without actually improving any sort of efficiency. It brings back bad memories of a second-year university assignment where we had to use C macros to reduce a program to the minimum possible number of lines, thereby making it completely incomprehensible 😝

I also find condensed lines like that _harder_ to read in general, since more is happening on a single line.  Too much time staring at ASM code where every line is a single operation, I suppose.

Anyway, I can bite that particular bullet for the sake of Python, I guess.

Reply:
I marked this one as 'optional' for exactly this reason. Some people just don't jive with list comprehensions, and that's fine. Feel free to change it back. 

I bring it up more as an educational point. When I see a `for` loop where a comprehension would do, it makes me think that there's something else going on that I'm missing. Some reason that the comprehension wasn't used, perhaps a side effect in the `for` loop, or something similar. This is why I say that I find the comprehension easier to read, it conveys intention a little better.

As for:
> reduce line numbers without actually improving any sort of efficiency

Traditionally list comprehensions are actually slightly more efficient than the equivalent `for` loop. The compiler was able to do some optimizations at the C level that it couldn't do with the standard `for`.  I say 'traditionally' because this is old CPython knowledge. I don't know if the same applies to micropython or not, and I don't know if they've been able to find a way to improve the `for` loop performance on par with the comprehension in more recent versions of CPython. I'll have to do some research.

All this is to say, there's plenty of reasons to use or not use comprehensions. It's as much of a style thing as anything. And you shouldn't feel the need to change your code because it makes it more readable to _me_.